File: native/sites/contribution_envelope.go
package sites

import (
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"sort"
)

func ParseContributionContent(input any) (map[string]any, error) {
	m, err := object(input, "type", "proposal")
	if err != nil || m["type"] != "site-contribution" {
		return nil, errors.New("conteúdo de proposta inválido")
	}
	cert, err := VerifyContribution(m["proposal"])
	if err != nil {
		return nil, err
	}
	return map[string]any{"type": "site-contribution", "proposal": cert}, nil
}

// MatchContributionEnvelope assumes an authenticated/decrypted outer envelope.
// It does not authorize inbox admission, owner approval, expiry or retransmission.
func MatchContributionEnvelope(bundle core.Bundle, plaintext any) (map[string]any, error) {
	content, err := ParseContributionContent(plaintext)
	if err != nil {
		return nil, err
	}
	cert := content["proposal"].(map[string]any)
	body := cert["body"].(map[string]any)
	author := body["contributor"].(map[string]any)["id"].(string)
	owner, _, _ := ParseAddress(body["target"].(map[string]any)["site"].(string))
	expected := []string{owner}
	if owner != author {
		expected = append(expected, author)
	}
	sort.Strings(expected)
	created, _ := contributionClock(body["created"])
	expires, _ := contributionClock(body["expires"])
	if bundle.Manifest.Kind != "site-contribution" || bundle.Manifest.Author.ID != author || bundle.Manifest.PublicKey != nil || !creationEqual(readersOf(bundle), expected) || bundle.Manifest.Created != created || bundle.Manifest.Expires != expires {
		return nil, errors.New("envelope diferente da proposta privada")
	}
	return content, nil
}

File: packages/sites/src/contribution-envelope.ts
import { canonical, exactShape, type Bundle } from "../../core/src/protocol";
import type { CertificateCrypto } from "../../core/src/certificate-types";
import { parseSiteAddress } from "./protocol";
import {
  createSiteContributionProtocol,
  type SiteContribution,
} from "./contribution-protocol";
export interface SiteContributionContent {
  type: "site-contribution";
  proposal: SiteContribution;
}
/** Certificate/envelope binding. The caller authenticates and decrypts the
 * outer bundle first and separately enforces expiry, policy and inbox limits. */
export function createContributionEnvelopeProtocol(crypto: CertificateCrypto) {
  const proposals = createSiteContributionProtocol(crypto);
  function content(input: unknown): SiteContributionContent {
    if (
      !exactShape(input, ["type", "proposal"]) ||
      (input as SiteContributionContent).type !== "site-contribution"
    )
      throw Error("Conteúdo de proposta inválido");
    return {
      type: "site-contribution",
      proposal: proposals.verify((input as SiteContributionContent).proposal),
    };
  }
  function match(bundle: Bundle, plaintext: unknown) {
    const value = content(plaintext),
      b = value.proposal.body,
      m = bundle.manifest,
      owner = parseSiteAddress(b.target.site).ownerId;
    const readers = [...new Set([owner, b.contributor.id])].sort();
    if (
      m.kind !== "site-contribution" ||
      m.author.id !== b.contributor.id ||
      m.publicKey !== null ||
      canonical(m.keys.map((k) => k.reader).sort()) !== canonical(readers) ||
      m.created !== b.created ||
      m.expires !== b.expires
    )
      throw Error("Envelope diferente da proposta privada");
    return value;
  }
  return { content, match };
}
